### DISCUSSION TITLE
타입스크립트의 union타입과 Intersection Type를 설명해주세요

### DISCUSSION THREAD

**Discussion by skyhanull** [wanted-9th-3team/frontend-study#35] [프론트 기술 질문 공간]
타입스크립트의 union타입과 Intersection Type를 각각 설명해주세요

**Comment by michoball** [comment]
union 타입은 |  기호를 사용하면 2개 이상의 타입중 하나일 수 있다는 것을 의미합니다. 

intersection 타입은 & 기호를 사용하며 & 로 연결된 타입을 다 가지는 하나의 새로운 타입을 의미합니다. 

ex) 
```typescript
function padLeft(value: string, padding: string | number) {
  // ...
}
// 여기서 padding은 string 일 수도 있고 number일 수도 있다는 의미로 union 타입을 사용함

 
interface ErrorHandling {
  success: boolean;
  error?: { message: string };
}
 
interface ArtworksData {
  artworks: { title: string }[];
}
type ArtworksResponse = ArtworksData & ErrorHandling;
// 이렇게하면 ArtworksResponse라는 타입은  다음과 같아진다. 

type  ArtworksResponse {
  success: boolean;
  error?: { message: string };
  artworks: { title: string }[];
}

```

**Comment by skyhanull** [comment]
union타입이란 | 기호를 사용하면 둘중 하나의 타입을 사용할 수 있는 타입입니다.
intersection타입은 &기호를 사용하여 타입을 넓힐 수 있습니다

이 두타입의 헷갈리는 차이점은 유니온 타입으로 | 기호를 사용한 타입은 교집합아닌 합집합이라는 점입니다.
유니온 타입을 사용 시 두 타입의 모든 key를 사용 할 수 있을 것 같지만 실제로는 공통적으로 있는 타입만 사용이 가능합니다.
반대로 intersection은 교집합과 같이 이용되어 두개의 타입의 속성을 모두 이용할 수 있게됩니다.

**Comment by bona373737** [comment]
Union Type 
- 자바스크립트의 OR 연산자(||)와 같이 'A' 이거나 'B'이다 라는 의미의 타입이다.
- 사용상의 주의점 : 
introduce() 함수를 호출하는 시점에 Person 타입이 올지 Developer 타입이 올지 알 수가 없기 때문에 어느 타입이 들어오든 간에 오류가 안 나는 방향으로 타입을 추론하게 된다. 때문에 기본적으로는 Person과 Developer 두 타입에 공통적으로 들어있는 속성인 name만 접근할 수 있게 된다.
interface Person {
  name: string;
  age: number;
}
interface Developer {
  name: string;
  skill: string;
}
function introduce(someone: Person | Developer) {
  someone.name; // O 정상 동작
  someone.age; // X 타입 오류
  someone.skill; // X 타입 오류
}

Intersection Type 
- 여러 타입을 모두 만족하는 하나의 타입을 의미한다.
- & 연산자를 이용해 여러 개의 타입 정의를 하나로 합치는 방식으로 사용가능하다. 
interface Person {
    name: string;
    age: number;
}
interface Developer {
    name: string;
    skill: number;
}
type Capt = Person & Developer;

--> Capt타입은 최종적으로 아래와 같이 정의된다.
{
  name: string;
  age: number;
  skill: string;
}

**Comment by WooGie911** [comment]
유니온 타입(Union Type)과 교차 타입(Intersection Type)은 타입을 구성할 수 있는 방법 중 하나입니다.

두가지 모두 이미 존재하는 타입을 구성하거나 결합하는 방법으로,

유니온 타입(Union Type)은 여러 타입 중 하나가 될 수 있는 값을 의미합니다.
세로 막대 (|)로 각 타입을 구분하며 자바스크립트의 OR 연산자(||)와 같이 'A' 이거나 'B'이다 라는 의미입니다.

교차 타입은 여러 타입을 하나로 결합합니다. 
기존 타입을 합쳐 필요한 기능을 모두 가진 단일 타입을 얻을 수 있습니다.
& 연산자를 이용해 여러 개의 타입 정의를 하나로 합치는 방식으로 'A' 이며 'B'이다 라는 의미입니다.
